### DISCUSSION TITLE
Problem setUp() test

### DISCUSSION THREAD

**Discussion by parishill25** [Cyfrin/foundry-full-course-cu#1191] [Q&A]
Hi,
I got this error while testing my testGetUsdValue() function:
[FAIL. Reason: setup failed: panic: array out-of-bounds access (0x32)] setUp() (gas: 0)
The application panicked (crashed).

//DscEngine.sol///
//SPDX-License-Identifier:MIT

pragma solidity ^0.8.18;

import {Ownable} from "@openzeppelin/contracts/access/Ownable.sol";
import {DecentralizedStableCoin} from "./DecentralizedStableCoin.sol";
import {ReentrancyGuard} from "@openzeppelin/contracts/security/ReentrancyGuard.sol";
import {IERC20} from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
import {AggregatorV3Interface} from "@chainlink/contracts/src/v0.8/interfaces/AggregatorV3Interface.sol";

contract DscEngine is ReentrancyGuard {
    // ERRORS//
    error DscEngine__LENGHTPROBLEM();
    error DscEngine__AmountShouldBeMoreThanZERO();
    error DSCEngine__NOTALLOWEDTOKEN();
    error DscEngine__TransferFailed();
    error DscEngine__HEALTHFACTORISBROKEN(uint256 userHealthFactor);
    error DecentralizedStableCoin__MINTEDFAILED();

    // STATE VARIABLES//
    uint256 private constant ADDITIONAL_FEED = 1e10;
    uint256 private constant PRECISION = 1e18;
    uint private constant LIQUIDATION_RATIO = 50;
    uint256 private constant LIQUIDATION_PRECISION = 100;
    uint256 private constant HEALTH_RATIO = 1;
    DecentralizedStableCoin private immutable i_Dsc;
    address[] private s_collateraltokens;

    mapping(address tokenAddress => address priceFeedAddress) s_priceFeed;
    mapping(address user => mapping(address tokenAddres => uint256 collateralAmount)) s_collateralDeposed;
    mapping(address user => uint256 amountUsdMinted) s_DscMinted;

    // EVENTS//
    event collateralDeposed(
        address indexed sender,
        address indexed tokenAddress,
        uint256 indexed amountCollateral
    );

    constructor(
        address[] memory tokenAddress,
        address[] memory priceFeedAddress,
        address DscAddress
    ) {
        if (tokenAddress.length != priceFeedAddress.length) {
            revert DscEngine__LENGHTPROBLEM();
        }

        for (uint256 i = 0; i <= tokenAddress.length; i++) {
            s_priceFeed[tokenAddress[i]] = priceFeedAddress[i];
            s_collateraltokens.push(tokenAddress[i]);
        }

        i_Dsc = DecentralizedStableCoin(DscAddress);
    }

    // Modifiers//
    modifier Morethanzero(uint256 amount) {
        if (amount == 0) {
            revert DscEngine__AmountShouldBeMoreThanZERO();
        }
        _;
    }

    modifier OnlyAllowedToken(address tokenAddress) {
        if (tokenAddress == address(0)) {
            revert DSCEngine__NOTALLOWEDTOKEN();
        }
        _;
    }

    // PUBLIC FUNCTIONS//

    function depositCollateral(
        address tokenAddress,
        uint256 amountCollateral
    )
        public
        Morethanzero(amountCollateral)
        OnlyAllowedToken(tokenAddress)
        nonReentrant
    {
        s_collateralDeposed[msg.sender][tokenAddress] += amountCollateral;
        emit collateralDeposed(msg.sender, tokenAddress, amountCollateral);
        bool success = IERC20(tokenAddress).transferFrom(
            msg.sender,
            address(this),
            amountCollateral
        );
        if (!success) {
            revert DscEngine__TransferFailed();
        }
    }

    function mintDsc(uint256 amountDscMinted) public {
        s_DscMinted[msg.sender] += amountDscMinted;

        _getHealthFunctionRevertedifIsBroken(msg.sender);

        bool minted = i_Dsc.mint(msg.sender, amountDscMinted);

        if (!minted) {
            revert DecentralizedStableCoin__MINTEDFAILED();
        }
    }

    function getAccountInformation(
        address user
    ) private view returns (uint256 DscMinted, uint256 amountCollateralInUsd) {
        DscMinted = s_DscMinted[user];
        amountCollateralInUsd = getCollateralInUsd(user);
    }

    function getHealthFactor(address user) private view returns (uint256) {
        (
            uint256 DscMinted,
            uint256 amountCollateralInUsd
        ) = getAccountInformation(user);

        uint256 colleralFactorRatio = (amountCollateralInUsd *
            LIQUIDATION_RATIO) / LIQUIDATION_PRECISION;

        uint256 healthFactor = (colleralFactorRatio * PRECISION) / DscMinted;

        return healthFactor;
    }

    function _getHealthFunctionRevertedifIsBroken(address user) internal view {
        uint256 userhealthFactor = getHealthFactor(user);

        if (userhealthFactor < HEALTH_RATIO) {
            revert DscEngine__HEALTHFACTORISBROKEN(userhealthFactor);
        }
    }

    function getCollateralInUsd(
        address user
    ) public view returns (uint256 totalAmountInUsd) {
        for (uint256 i = 0; i <= s_collateraltokens.length; i++) {
            address token = s_collateraltokens[i];
            uint256 amount = s_collateralDeposed[user][token];
            totalAmountInUsd += getUsdValue(token, amount);
        }
    }

    function getUsdValue(
        address tokenAddress,
        uint256 amount
    ) public view returns (uint256) {
        AggregatorV3Interface priceFeed = AggregatorV3Interface(
            s_priceFeed[tokenAddress]
        );
        (, int256 price, , , ) = priceFeed.latestRoundData();
        return ((uint256(price) * ADDITIONAL_FEED) * amount) / PRECISION;
    }
}

//////DecentralizedStableCoin.sol////:
//SPDX-License-Identifier:MIT

pragma solidity ^0.8.18;

import {ERC20Burnable, ERC20} from "@openzeppelin/contracts/token/ERC20/extensions/ERC20Burnable.sol";
import {Ownable} from "@openzeppelin/contracts/access/Ownable.sol";

contract DecentralizedStableCoin is ERC20Burnable, Ownable {
    error DecentralizedSTableCoin__NEEDMORECOINS();
    error DecentralizedSTableCoin__BALANCENOTENOUGH();
    error DecentralizedStableCoin__NOTRIGHTADDRESS();

    uint256 private s_tokenCounter;

    constructor() ERC20("DecentralizedStableCoin", "DSC") {}

    function burn(uint256 _amount) public override onlyOwner {
        uint256 balance = balanceOf(msg.sender);
        if (_amount <= 0) {
            revert DecentralizedSTableCoin__NEEDMORECOINS();
        }

        if (balance < _amount) {
            revert DecentralizedSTableCoin__BALANCENOTENOUGH();
        }

        super.burn(_amount);
    }

    function mint(address _to, uint256 _amount) public returns (bool) {
        if (_to == address(0)) {
            revert DecentralizedStableCoin__NOTRIGHTADDRESS();
        }

        if (_amount <= 0) {
            revert DecentralizedSTableCoin__NEEDMORECOINS();
        }
        _mint(_to, _amount);
        return true;
    }
}
////DeployDsc.sol////
//SPDX-License-Identifier:MIT

pragma solidity ^0.8.18;

import {Script, console} from "forge-std/Script.sol";
import {DscEngine} from "src/DscEngine.sol";
import {DecentralizedStableCoin} from "src/DecentralizedStableCoin.sol";
import {HelperConfig} from "./HelperConfig.sol";

contract DeployDsc is Script {
    address[] tokenAddresses;
    address[] priceFeedAddresses;
    DscEngine engine;
    DecentralizedStableCoin dscContract;

    function run()
        public
        returns (DecentralizedStableCoin, DscEngine, HelperConfig)
    {
        HelperConfig helperConfig = new HelperConfig();
        (
            address wethUsdPriceFeed,
            address wbtcUsdPriceFeed,
            address weth,
            address wbtc,
            uint256 deployerKey
        ) = helperConfig.activeNetwork();

        tokenAddresses = [weth, wbtc];
        priceFeedAddresses = [wethUsdPriceFeed, wbtcUsdPriceFeed];

        vm.startBroadcast(deployerKey);
        dscContract = new DecentralizedStableCoin();
        engine = new DscEngine(
            tokenAddresses,
            priceFeedAddresses,
            address(dscContract)
        );
        vm.stopBroadcast();
        dscContract.transferOwnership(address(engine));
        return (dscContract, engine, helperConfig);
    }
}

////HELPERCONFIG///
//SPDX-License-Identifier:MIT

pragma solidity ^0.8.18;

import {MockV3Aggregator} from "../test/mocks/MockV3Aggregator.sol";
import {Script, console} from "forge-std/Script.sol";
import {ERC20Mock} from "@openzeppelin/contracts/mocks/ERC20Mock.sol";

contract HelperConfig is Script {
    NetworkConfig public activeNetwork;
    uint8 public constant DECIMALS = 8;
    int256 public constant ETH_USD_PRICE = 2000e8;
    int256 public constant BTC_USD_PRICE = 1000e8;

    uint256 public DEFAULT_ANVIL_PRIVATE_KEY =
        0xac0974bec39a17e36ba4a6b4d238ff944bacb478cbed5efcae784d7bf4f2ff80;

    constructor() {
        if (block.chainid == 11155111) {
            activeNetwork = getSepoliaConfigNetwork();
        } else {
            activeNetwork = getAnvilConfigNetwork();
        }
    }

    struct NetworkConfig {
        address wethUsdPriceFeed;
        address wbtcUsdPriceFeed;
        address weth;
        address wbtc;
        uint256 deployerKey;
    }

    function getSepoliaConfigNetwork()
        public
        view
        returns (NetworkConfig memory)
    {
        return
            NetworkConfig({
                wethUsdPriceFeed: 0x694AA1769357215DE4FAC081bf1f309aDC325306,
                wbtcUsdPriceFeed: 0x1b44F3514812d835EB1BDB0acB33d3fA3351Ee43,
                weth: 0xdd13E55209Fd76AfE204dBda4007C227904f0a81,
                wbtc: 0x8f3Cf7ad23Cd3CaDbD9735AFf958023239c6A063,
                deployerKey: vm.envUint("PRIVATE_KEY")
            });
    }

    function getAnvilConfigNetwork() public returns (NetworkConfig memory) {
        if (activeNetwork.wethUsdPriceFeed != address(0)) {
            return activeNetwork;
        }

        vm.startBroadcast();
        MockV3Aggregator ethUsdPriceFeed = new MockV3Aggregator(
            DECIMALS,
            ETH_USD_PRICE
        );
        ERC20Mock wethMock = new ERC20Mock("WETH", "WETH", msg.sender, 1000e8);

        MockV3Aggregator btcUsdPriceFeed = new MockV3Aggregator(
            DECIMALS,
            BTC_USD_PRICE
        );
        ERC20Mock wbtcMock = new ERC20Mock("WBTC", "WBTC", msg.sender, 1000e8);
        vm.stopBroadcast();

        return
            NetworkConfig({
                wethUsdPriceFeed: address(ethUsdPriceFeed),
                weth: address(wethMock),
                wbtcUsdPriceFeed: address(btcUsdPriceFeed),
                wbtc: address(wbtcMock),
                deployerKey: DEFAULT_ANVIL_PRIVATE_KEY
            });
    }
}
////Unit Test/////
//SPDX-License-Identifier:MIT

pragma solidity ^0.8.18;

import {Test} from "forge-std/Test.sol";
import {DeployDsc} from "script/DeployDsc.sol";
import {DscEngine} from "src/DscEngine.sol";
import {DecentralizedStableCoin} from "src/DecentralizedStableCoin.sol";
import {HelperConfig} from "script/HelperConfig.sol";

contract DscEngineTest is Test {
    DecentralizedStableCoin dsccontract;
    DscEngine dscengine;
    DeployDsc deployer;
    HelperConfig helperconfig;
    address ethUsdPriceFeed;
    address weth;

    function setUp() public {
        deployer = new DeployDsc();
        (dsccontract, dscengine, helperconfig) = deployer.run();
        (ethUsdPriceFeed, , weth, , ) = helperconfig.activeNetwork();
    }

    // TestPriceFeed//
    function testGetUsdValue() public {
        uint256 ethAmount = 15e18;
        uint256 usdexpected = 30000e18;
        uint256 actualusd = dscengine.getUsdValue(weth, ethAmount);
        assertEq(usdexpected, actualusd);
    }
}

**Comment by xyz899** [comment]
Hi,
could you rerun your test using -vvvvv to see the tree logs  and paste it here ?
